File: mobile-engage/src/main/java/com/emarsys/mobileengage/event/DefaultEventServiceInternal.java
package com.emarsys.mobileengage.event;

import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.request.RequestManager;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.request.MobileEngageRequestModelFactory;

import java.util.Map;

public class DefaultEventServiceInternal implements EventServiceInternal {
    private final MobileEngageRequestModelFactory requestModelFactory;
    private final RequestManager requestManager;

    public DefaultEventServiceInternal(RequestManager requestManager, MobileEngageRequestModelFactory requestModelFactory) {
        Assert.notNull(requestModelFactory, "RequestModelFactory must not be null!");
        Assert.notNull(requestManager, "RequestManager must not be null!");

        this.requestModelFactory = requestModelFactory;
        this.requestManager = requestManager;
    }

    @Override
    public String trackCustomEvent(String eventName, Map<String, String> eventAttributes, CompletionListener completionListener) {
        Assert.notNull(eventName, "EventName must not be null!");

        RequestModel requestModel = requestModelFactory.createCustomEventRequest(eventName, eventAttributes);
        requestManager.submit(requestModel, completionListener);

        return requestModel.getId();
    }

    @Override
    public String trackInternalCustomEvent(String eventName, Map<String, String> eventAttributes, CompletionListener completionListener) {
        Assert.notNull(eventName, "EventName must not be null!");

        RequestModel requestModel = requestModelFactory.createInternalCustomEventRequest(eventName, eventAttributes);

        requestManager.submit(requestModel, completionListener);

        return requestModel.getId();
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/event/DefaultEventServiceInternal.kt
package com.emarsys.mobileengage.event

import com.emarsys.core.Mockable
import com.emarsys.core.api.result.CompletionListener
import com.emarsys.core.request.RequestManager
import com.emarsys.core.util.Assert
import com.emarsys.mobileengage.request.MobileEngageRequestModelFactory

@Mockable
class DefaultEventServiceInternal(
    private val requestModelFactory: MobileEngageRequestModelFactory,
    private val requestManager: RequestManager
) : EventServiceInternal {

    override fun trackCustomEvent(
        eventName: String,
        eventAttributes: Map<String, String>?,
        completionListener: CompletionListener?
    ): String? {
        Assert.notNull(eventName, "EventName must not be null!")
        val requestId = try {
            val requestModel =
                requestModelFactory.createCustomEventRequest(eventName, eventAttributes)
            requestManager.submit(requestModel, completionListener)
            requestModel.id
        } catch (e: IllegalArgumentException) {
            completionListener?.onCompleted(e)
            null
        }
        return requestId
    }

    override fun trackCustomEventAsync(
        eventName: String,
        eventAttributes: Map<String, String>?,
        completionListener: CompletionListener?
    ) {
        trackCustomEvent(eventName, eventAttributes, completionListener)
    }

    override fun trackInternalCustomEvent(
        eventName: String,
        eventAttributes: Map<String, String>?,
        completionListener: CompletionListener?
    ): String? {
        Assert.notNull(eventName, "EventName must not be null!")
        val requestId = try {
            val requestModel =
                requestModelFactory.createInternalCustomEventRequest(eventName, eventAttributes)
            requestManager.submit(requestModel, completionListener)
            requestModel.id
        } catch (e: IllegalArgumentException) {
            completionListener?.onCompleted(e)
            null
        }
        return requestId
    }

    override fun trackInternalCustomEventAsync(
        eventName: String,
        eventAttributes: Map<String, String>?,
        completionListener: CompletionListener?
    ) {
        trackInternalCustomEvent(eventName, eventAttributes, completionListener)
    }
}